### PR TITLE
Allow gradient transforms applied to broadcasted tapes if the broadcasted parameter(s) are not trainable

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
In #4462 we discussed the differentiation of broadcasted tapes/QNodes with gradient transforms. Due to some technical debt in the code for JVPs and due to unclear desired behaviour, applying gradient transforms to such QNodes was disallowed in #4480.
However, this also disallows applying gradient transforms to broadcasted tapes for which the broadcasted parameter(s) is/are not trainable, even though the problems discussed in #4462 do not apply in this case.

**Description of the Change:**
This PR lifts the restriction of applying gradient transforms to broadcasted tapes for the case that no trainable parameters are broadcasted.
The problem described in [this forum post](https://discuss.pennylane.ai/t/qml-training-using-shot-based-simulator/4206) can be resolved by this, for example.

**Benefits:**
Unlocks gradient transforms for broadcasted tapes/QNodes as long as the broadcasting is in non-trainable parameters (which is quite common for QML applications: You want to broadcast across data points rather than trainable parameters.).

**Possible Drawbacks:**
Usage of `tape.trainable_params` although we don't want to rely on this property too much in the future.

### CODE REVIEW

File: pennylane/gradients/gradient_transform.py [hunk 1]
```diff
@@ -99,18 +99,26 @@ def assert_no_variance(measurements, transform_name):
...
     if tape.batch_size is not None:
-        raise NotImplementedError(
-            f"Computing the gradient of broadcasted tapes with the {transform_name} "
-            "gradient transform is currently not supported. See #4462 for details."
-        )
+        trainable_batch_size = None
+        for idx in range(len(tape.trainable_params)):
+            op_batch_size = tape.get_operation(idx)[0].batch_size
+            if op_batch_size is not None:
+                trainable_batch_size = op_batch_size
```
Comment (albi3ro) [inline pennylane/gradients/gradient_transform.py:114]: Couldn't we just raise the error here? Is there a reason to defer raising the error to later?

Comment (dwierichs) [reply pennylane/gradients/gradient_transform.py]: Simplified the function.

Comment (albi3ro) [reply pennylane/gradients/gradient_transform.py]: 👍

File: tests/gradients/core/test_hadamard_gradient.py [hunk 2]
```diff
@@ -66,13 +66,34 @@ def cost6(x):
...
+        parameter is not differentiated, and that the results correspond to the stacked
+        results of the single-tape derivatives."""
+        dev = qml.device("default.qubit")
+        x = [0.4, 0.2]
+        tape = qml.tape.QuantumScript([qml.RY(0.6, 0), qml.RX(x, 0)], [qml.expval(qml.PauliZ(0))])
+        tape.trainable_params = [0]
+        batched_tapes, batched_fn = qml.gradients.hadamard_grad(tape)
+        batched_grad = batched_fn(dev.execute(batched_tapes))
+        separate_tapes = [
+            qml.tape.QuantumScript([qml.RY(0.6, 0), qml.RX(_x, 0)], [qml.expval(qml.PauliZ(0))])
```
Comment (albi3ro) [inline tests/gradients/core/test_hadamard_gradient.py:88]:
```suggestion
            qml.tape.QuantumScript([qml.RY(0.6, 0), qml.RX(_x, 0)], [qml.expval(qml.PauliZ(0))], trainable_params=[0])
```

Can now be set on initialization, so you don't have to come back and update it.

**Review by albi3ro — APPROVED** [review]
👍

File: tests/gradients/parameter_shift/test_parameter_shift.py [hunk 3]
```diff
@@ -902,6 +906,32 @@ def test_with_multiple_parameters_broadcasted(self, dim, argnum):
...
+        y = -0.654
+        if pos == 1:
+            x, y = y, x
+
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.RX(x, wires=[0])
+            qml.RY(y, wires=[0])  # does not have any impact on the expval
+            qml.expval(qml.PauliZ(0))
+
+        tape = qml.tape.QuantumScript.from_queue(q, trainable_params=[1 - pos])
```
Comment (dwierichs) [inline tests/gradients/parameter_shift/test_parameter_shift.py:925]:
```suggestion
        tape = qml.tape.QuantumScript.from_queue(q)
        tape.trainable_params = [1 - pos]
```

**Review by vincentmr — APPROVED** [review]
Hi @dwierichs , looks good to me :)

**Pull request merged**